streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```
@@ -38,346 +40,644 @@
+    // possible that when newly created tasks are created the thread and its changelog reader are still restoring existing
+    // active tasks, and hence this function is idempotent and can be called multiple times.
+    //
+    // NOTE: even if the newly created tasks do not need any restoring, we still first transit to this state and then
+    // immediately transit back -- there's no overhead of transiting back and forth but simplifies the logic a lot.
+    // TODO K9113: this function should be called by stream thread
+    public void transitToRestoreActive() {
+        log.debug("Transiting to restore active tasks: {}", changelogs);
 
-        needsInitializing.add(restorer.partition());
+        // pause all partitions that are for standby tasks from the restore consumer
+        pauseChangelogsFromRestoreConsumer(standbyRestoringChangelogs());
+
+        state = ChangelogReaderState.ACTIVE_RESTORING;
     }
 
-    public Collection<TopicPartition> restore(final RestoringTasks active) {
-        if (!needsInitializing.isEmpty()) {
-            initialize(active);
+    // Only after we've completed restoring all active tasks we'll then move back to resume updating standby tasks.
+    // This function is NOT idempotent: if it is already in updating standby tasks mode, we should not call it again.
+    //
+    // NOTE: we do not clear completed active restoring changelogs or remove partitions from restore consumer either
+    // upon completing them but only pause the corresponding partitions; the changelog metadata / partitions would only
+    // be cleared when the corresponding task is being removed from the thread.
+    // TODO K9113: this function should be called by stream thread
+    public void transitToUpdateStandby() {
+        if (state != ChangelogReaderState.ACTIVE_RESTORING) {
+            throw new IllegalStateException("The changelog reader is not restoring actove tasks while trying to " +
```

```suggestion
            throw new IllegalStateException("The changelog reader is not restoring active tasks while trying to " +
```

---

prop: also log an error here, maybe even include `changelogs` (or at least `changelogs.keyset`?)

---

I will include the changelogs in the exception message.

Since this is a fatal error, the illegal state exception would never be swallowed and will always be printed to std. So I feel logging another error would have duplicated entries on log files.